From d412c7630978f888e8db85bec9f39e265d24b006 Mon Sep 17 00:00:00 2001
From: Xiaoliang Yang <xiaoliang.yang_1@nxp.com>
Date: Wed, 13 Mar 2019 15:09:44 +0800
Subject: [PATCH 663/706] mscc: switch: fix cbstreamidset issue

1. Use mac learn to insert mac data into mac table.
2. Modify 802.1cb recover set bug.

Signed-off-by: Xiaoliang Yang <xiaoliang.yang_1@nxp.com>
(cherry picked from commit 84b13de7cee5429d06184e037c761b02c62a99e7)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/mscc/tsn_switch.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/mscc/tsn_switch.c b/drivers/net/ethernet/mscc/tsn_switch.c
index c28c9247ff73..4a433460a2e0 100644
--- a/drivers/net/ethernet/mscc/tsn_switch.c
+++ b/drivers/net/ethernet/mscc/tsn_switch.c
@@ -593,6 +593,12 @@ int switch_cb_streamid_set(struct net_device *ndev, u32 index, bool enable,
 			     ANA_TABLES_MACACCESS_DEST_IDX(dst_idx) |
 			     ANA_TABLES_MACACCESS_MAC_TABLE_CMD(MACACCESS_CMD_WRITE),
 			     ANA_TABLES_MACACCESS);
+
+		ocelot_write(ocelot, ANA_TABLES_MACACCESS_VALID |
+			     ANA_TABLES_MACACCESS_ENTRYTYPE(1) |
+			     ANA_TABLES_MACACCESS_DEST_IDX(dst_idx) |
+			     ANA_TABLES_MACACCESS_MAC_TABLE_CMD(MACACCESS_CMD_LEARN),
+			     ANA_TABLES_MACACCESS);
 	} else {
 		netdev_info(ndev, "disable stream set\n");
 		mac = streamid->para.nid.dmac;
@@ -1060,7 +1066,8 @@ int switch_seq_rec_set(struct net_device *ndev, u32 index,
 		   sr_conf->seq_len, sr_conf->his_len,
 		   sr_conf->rtag_pop_en);
 
-	ocelot_rmw_rix(ocelot, 1, ANA_PORT_MODE_REDTAG_PARSE_CFG,
+	ocelot_rmw_rix(ocelot, ANA_PORT_MODE_REDTAG_PARSE_CFG,
+		       ANA_PORT_MODE_REDTAG_PARSE_CFG,
 		       ANA_PORT_MODE, port->chip_port);
 
 	ocelot_write(ocelot,
-- 
2.17.1

